### ISSUE TITLE
MemcacheTest inconsistent

### ISSUE THREAD

**Issue by shopblocks** [UnionOfRAD/lithium#1230]
Locally, the memcache tests are inconsistent in their passing.

To me, it feels as though it may be an issue with the delay between getting something written into memcache.

The current code is this:

```php
$keys = array('key1' => 'data1');
$expiry = '+1 second';
$this->memcache->write($keys, $expiry);

sleep(2);

$result = $this->_conn->get('key1');
$this->assertFalse($result);
```

This sometimes passes, and sometimes doesn't.

If I increase the sound to 5 seconds, it always passes.

2 second sleep:
```
1450278957.1978
1450278957.198
1450278959.1983
string(5) "data1"
```

5 second sleep:

```
1450278985.6084
1450278985.6088
1450278990.609
bool(false)
```

I'm not sure if there is perhaps an underlying issue with how caching is done, but the tests being inconsistent is worrying.

**Comment by davidpersson** [comment]
In general it's good practice to make tests time-insensitive. In this case however, time is actually what we want to test. For me the tests seem to run fine ("always").

We don't want to increase sleep delay here, as this will slow down test execution. 1s increase is OK in my opinion. BUT before we do so please verify that:

- when running just the single MemcacheTest you still see the fails -- otherwise this might be an issue with other tests leaking into this one
- if increasing the delay by 1s only the problem is fixed

**Comment by shopblocks** [comment]
Please can you clarify how to run just one of those particular tests? I cannot see how to do it in the documentation.

**Comment by davidpersson** [comment]
You can pass a path to a file or directory to the test command `console/li3 test tests/cases/storage/cache/adapter/MemcacheTest.php`

**Comment by shopblocks** [comment]
It does not always pass when running just that file.

`bash console/li3 test tests/integration/storage/cache/adapter/MemcacheTest.php --verbose --no-color |& tee /var/www/vhosts/lithium/output.txt`

```
----
Test
----

[     pass] on line   49 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testEnabled()
[     pass] on line   60 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testSimpleWrite()
[     pass] on line   61 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testSimpleWrite()
[     pass] on line   64 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testSimpleWrite()
[     pass] on line   74 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testSimpleWrite()
[     pass] on line   78 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testSimpleWrite()
[     pass] on line   81 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testSimpleWrite()
[     pass] on line   92 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testWriteExpiryDefault()
[     pass] on line   96 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testWriteExpiryDefault()
[     pass] on line   99 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testWriteExpiryDefault()
[     pass] on line  109 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testWriteNoExpiry()
[     pass] on line  112 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testWriteNoExpiry()
[     pass] on line  120 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testWriteNoExpiry()
[     pass] on line  123 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testWriteNoExpiry()
[     pass] on line  131 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testWriteNoExpiry()
[     pass] on line  134 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testWriteNoExpiry()
[     pass] on line  145 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testWriteExpiryExpires()
[     fail] on line  156 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testWriteExpiryExpires()


FAIL

17 / 18 passes
1 fail and 0 exceptions

Failed assertion assertFalse.
 File    : /var/www/vhosts/lithium/tests/integration/storage/cache/adapter/MemcacheTest.php
 Class   : lithium\tests\integration\storage\cache\adapter\MemcacheTest
 Method  : testWriteExpiryExpires()
 Line    : 156
 ________
expected: false
result: 'data1'

 ________

```

**Comment by shopblocks** [comment]
It also fails if i comment out all tests except that one.

```
----
Test
----

[     pass] on line  145 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testWriteExpiryExpires()
[     fail] on line  156 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testWriteExpiryExpires()


FAIL

1 / 2 pass
1 fail and 0 exceptions

Failed assertion assertFalse.
 File    : /var/www/vhosts/lithium/tests/integration/storage/cache/adapter/MemcacheTest.php
 Class   : lithium\tests\integration\storage\cache\adapter\MemcacheTest
 Method  : testWriteExpiryExpires()
 Line    : 156
 ________
expected: false
result: 'data1'

 ________
```

**Comment by shopblocks** [comment]
Increasing the timeout to 3 seconds appears to make it pass consistently when ran on its own.

**Comment by shopblocks** [comment]
The same issue also happens on the testWriteExpiryTtl function, that appears to d othe same thing, but where expiry is a number not a string.

```php
$keys = array('key1' => 'data1');
$expiry = 1;
$this->memcache->write($keys, $expiry);
```

Increasing this to 3 seconds also fixes that test.

**Comment by shopblocks** [comment]
I managed to get it to fail running also just one test case on the 2 seconds.

```php
	public function testWriteExpiryExpires() {
//		$keys = array('key1' => 'data1');
//		$expiry = '+5 seconds';
//		$this->memcache->write($keys, $expiry);
//
//		$result = (boolean) $this->_conn->get('key1');
//		$this->assertTrue($result);
//
//		$this->_conn->delete('key1');

		$keys = array('key1' => 'data1');
		$expiry = '+1 second';
		$this->memcache->write($keys, $expiry);

		sleep(2);

		$result = $this->_conn->get('key1');
		$this->assertFalse($result);
	}
```

```
----
Test
----

[     fail] on line  156 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testWriteExpiryExpires()


FAIL

0 / 1 passes
1 fail and 0 exceptions

Failed assertion assertFalse.
 File    : /var/www/vhosts/lithium/tests/integration/storage/cache/adapter/MemcacheTest.php
 Class   : lithium\tests\integration\storage\cache\adapter\MemcacheTest
 Method  : testWriteExpiryExpires()
 Line    : 156
 ________
expected: false
result: 'data1'

 ________
```

**Comment by shopblocks** [comment]
The memcached version I have installed is `VERSION 1.4.14 (Ubuntu)`

**Comment by shopblocks** [comment]
This fails for me locally on both PHP 7.0 and PHP 5.6.14

**Comment by shopblocks** [comment]
It has also failed locally for me on a 3 second delay. I think there must be a different underlying issue to this.

@davidpersson 

```php
	public function testWriteExpiryExpires() {
//		$keys = array('key1' => 'data1');
//		$expiry = '+5 seconds';
//		$this->memcache->write($keys, $expiry);
//
//		$result = (boolean) $this->_conn->get('key1');
//		$this->assertTrue($result);
//
//		$this->_conn->delete('key1');

		$keys = array('key1' => 'data1');
		$expiry = '+1 second';
		$this->memcache->write($keys, $expiry);

		sleep(3);

		$result = $this->_conn->get('key1');
		$this->assertFalse($result);
	}
```

```
----
Test
----

[     fail] on line  156 in lithium\tests\integration\storage\cache\adapter\MemcacheTest::testWriteExpiryExpires()


FAIL

0 / 1 passes
1 fail and 0 exceptions

Failed assertion assertFalse.
 File    : /var/www/vhosts/lithium/tests/integration/storage/cache/adapter/MemcacheTest.php
 Class   : lithium\tests\integration\storage\cache\adapter\MemcacheTest
 Method  : testWriteExpiryExpires()
 Line    : 156
 ________
expected: false
result: 'data1'

 ________
```

This was on PHP 5.6.14

**Comment by shopblocks** [comment]
PHP Memcached information for PHP 5.6.14 
![image](https://cloud.githubusercontent.com/assets/14819903/11866238/ac49931c-a4a1-11e5-8898-3ff3e7fa32d1.png)

PHP Memcached Information for PHP 7.0.0
![image](https://cloud.githubusercontent.com/assets/14819903/11866254/bfb968f0-a4a1-11e5-88be-4a67b3e80d01.png)

**Comment by davidpersson** [comment]
Okay, thanks for all the helpful information. I think this behavior is due to internals of memcached and how it determines stale items. I've tried to simplify the test in e1fce7c1ba4eb8f9ee915b74017c9147a432dc71. If this doesn't help, we need to decide if raising the limit to 3 or 4s, fixes the issue. Otherwise we will need to allow the test to skip when it operates outside boundaries.

**Comment by shopblocks** [comment]
This fails all the time now.

![image](https://cloud.githubusercontent.com/assets/14819903/11870328/a17b2e92-a4be-11e5-93bc-1faabe53feba.png)

![image](https://cloud.githubusercontent.com/assets/14819903/11870339/b2e4754e-a4be-11e5-95ad-78af9ffc69e4.png)

**Comment by davidpersson** [comment]
At least it's not random anymore ;)

**Comment by shopblocks** [comment]
Turning it to 3 seconds does not always pass, but does *more often*

**Comment by shopblocks** [comment]
5 seconds seems to always pass. Testing 4 seconds now.

**Comment by shopblocks** [comment]
4 seconds still fails. Testing 5 again now.

This is the command I use to test it:

```bash
while bash console/li3 test tests/integration/storage/cache/adapter/MemcacheTest.php ; do :; done
```

**Comment by shopblocks** [comment]
It has just passed running through 50 times successfully.

5 seconds seems to be where it works.

Not sure why this is, must be an issue with memcached and the vagrant box writing speed / delay.

**Comment by shopblocks** [comment]
@nateabele @davidpersson Any objection to upping the sleep to 5 seconds?

It is 1 second more than previously required (2 separate 2 second sleeps)

**Comment by nateabele** [comment]
Seems fine to me.

**Comment by shopblocks** [comment]
Created PR https://github.com/UnionOfRAD/lithium/pull/1237

**Comment by davidpersson** [comment]
Fixed by #1237
